archivist/sboms.py

```diff
@@ -0,0 +1,164 @@
+from .dictmerge import _deepmerge
+from .sbommetadata import SBOM
+
+#: Default page size - number of entities fetched in one REST GET in the
+#: :func:`~_AssetsClient.list` method. This can be overridden but should rarely
+#: be changed.
+DEFAULT_PAGE_SIZE = 500
+
+LOGGER = logging.getLogger(__name__)
+
+FIXTURE_LABEL = "sboms"
+
+
+class _SBOMSClient:
+    """SBOMSClient
+
+    Access to SBOMs entities using CRUD interface. This class is usually
+    accessed as an attribute of the Archivist class.
+
+    Args:
+        archivist (Archivist): :class:`Archivist` instance
+
+    """
+
+    def __init__(self, archivist: "type_helper.Archivist"):
+        self._archivist = archivist
+
+    def upload(self, fd: BinaryIO, *, mtype: str = "text/xml") -> SBOM:
+        """Create SBOM
+
+        Creates SBOM from opened file or other data source.
+
+        Args:
+            fd (file): opened file descriptor or other file-type iterable.
+            mtype (str): mimetype of data.
+
+        Returns:
+            :class:`SBOM` instance
+
+        """
+
+        LOGGER.debug("Upload SBOM")
+        return SBOM(
+            **self._archivist.post_file(
+                f"{SBOMS_SUBPATH}/{SBOMS_LABEL}",
+                fd,
+                mtype,
+                form="sbom",
+            )
+        )
+
+    def download(self, identity: str, fd: BinaryIO) -> Response:
+        """Read SBOM
+
+        Reads SBOM into data sink (usually a file opened for write)..
+        Note that returns the response as the body will be consumed by the
+        fd iterator
+
+        Args:
+            identity (str): SBOM identity e.g. sboms/xxxxxxxxxxxxxxxxxxxxxxx
+            fd (file): opened file escriptor or other file-type sink.
+
+        Returns:
+            REST response
+
+        """
+        return self._archivist.get_file(
+            SBOMS_SUBPATH,
+            identity,
+            fd,
+        )
+
+    def read(self, identity: str) -> SBOM:
+        """Read SBOM metadata
+
+        Reads SBOM metadata.
+
+        Args:
+            identity (str): sbom identity e.g. sboms/xxxxxxxxxxxxxxxxxxxxxxx
+
+        Returns:
+            BOM
+
+        """
+        return SBOM(
+            **self._archivist.get(
+                SBOMS_SUBPATH,
+                identity,
+                tail=SBOMS_METADATA,
+            )
+        )
+
+    def __query(self, metadata: Optional[Dict]) -> Dict:
+        query = deepcopy(metadata) if metadata else {}
+        return _deepmerge(self._archivist.fixtures.get(FIXTURE_LABEL), query)
+
+    def list(
+        self,
+        *,
+        page_size: int = DEFAULT_PAGE_SIZE,
+        metadata: Optional[Dict] = None,
+    ):
+        """List SBOMS.
+
+        Lists SBOMS that match criteria:
+
+            sboms = self.arch.sboms.list(metadata={"uploaded_since": "2021-11-16T11:28:31Z"})
+
+        Args:
+            metadata (dict): optional e.g. {"life_cycle_status": "ACTIVE" }
```

Comment:
Minor: maybe we need a complete example of additional things we can add to the dict, stuff like `trusted`, `uploaded_since`, `uploaded_before`, `privacy`, `personal` with explanations for what each does?

Reply:
I have added some - I will add more when doing withdraw/publish